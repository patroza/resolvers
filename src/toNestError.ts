import {
  set,
  get,
  FieldError,
  FieldErrors,
  Field,
  ResolverOptions,
} from 'react-hook-form';

export const toNestError = <TFieldValues>(
  errors: Record<string, FieldError>,
  options: ResolverOptions<TFieldValues>,
): FieldErrors<TFieldValues> => {
  const fieldErrors: FieldErrors<TFieldValues> = {};
  for (const path in errors) {
    const field = get(options.fields, path) as Field['_f'] | undefined;

    // Native validation (web only)
    if (
      options.shouldUseNativeValidation &&
      field &&
      'reportValidity' in field.ref
    ) {
      field.ref.setCustomValidity(errors[path].message || '');
      field.ref.reportValidity();
    }

    set(
      fieldErrors,
      path,
      Object.assign(errors[path], { ref: field && field.ref }),
    );
  }

  return fieldErrors;
};
