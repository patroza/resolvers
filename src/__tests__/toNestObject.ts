/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Field, FieldError, InternalFieldName } from 'react-hook-form';
import { toNestError } from '../toNestError';

const flatObject: Record<string, FieldError> = {
  name: { type: 'st', message: 'first message' },
  'test.0.name': { type: 'nd', message: 'second message' },
  'n.test': { type: 'rd', message: 'third message' },
};

const fields = {
  name: {
    ref: {
      reportValidity: jest.fn(),
      setCustomValidity: jest.fn(),
    },
  },
  n: {
    test: {
      ref: {
        reportValidity: jest.fn(),
        setCustomValidity: jest.fn(),
      },
    },
  },
  unused: {
    ref: { name: 'unusedRef' },
  },
} as any as Record<InternalFieldName, Field['_f']>;

test('transforms flat object to nested object', () => {
  expect(
    toNestError(flatObject, { fields, shouldUseNativeValidation: false }),
  ).toMatchSnapshot();
});

test('transforms flat object to nested object and shouldUseNativeValidation: true', () => {
  expect(
    toNestError(flatObject, { fields, shouldUseNativeValidation: true }),
  ).toMatchSnapshot();
  expect(
    (fields.name.ref as HTMLInputElement).reportValidity,
  ).toHaveBeenCalledTimes(1);
  expect(
    (fields.name.ref as HTMLInputElement).setCustomValidity,
  ).toHaveBeenCalledTimes(1);
  expect(
    (fields.name.ref as HTMLInputElement).setCustomValidity,
  ).toHaveBeenCalledWith(flatObject.name.message);

  // @ts-expect-error
  expect(fields.n?.test.ref.reportValidity).toHaveBeenCalledTimes(1);
  // @ts-expect-error
  expect(fields.n.test.ref.setCustomValidity).toHaveBeenCalledTimes(1);
  // @ts-expect-error
  expect(fields.n.test.ref.setCustomValidity).toHaveBeenCalledWith(
    flatObject['n.test'].message,
  );
});
