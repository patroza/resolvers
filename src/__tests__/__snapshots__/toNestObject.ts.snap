// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms flat object to nested object 1`] = `
Object {
  "n": Object {
    "test": Object {
      "message": "third message",
      "ref": Object {
        "reportValidity": [MockFunction],
        "setCustomValidity": [MockFunction],
      },
      "type": "rd",
    },
  },
  "name": Object {
    "message": "first message",
    "ref": Object {
      "reportValidity": [MockFunction],
      "setCustomValidity": [MockFunction],
    },
    "type": "st",
  },
  "test": Array [
    Object {
      "name": Object {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;

exports[`transforms flat object to nested object and shouldUseNativeValidation: true 1`] = `
Object {
  "n": Object {
    "test": Object {
      "message": "third message",
      "ref": Object {
        "reportValidity": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "setCustomValidity": [MockFunction] {
          "calls": Array [
            Array [
              "third message",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "type": "rd",
    },
  },
  "name": Object {
    "message": "first message",
    "ref": Object {
      "reportValidity": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "setCustomValidity": [MockFunction] {
        "calls": Array [
          Array [
            "first message",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    },
    "type": "st",
  },
  "test": Array [
    Object {
      "name": Object {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;
